component/metrics/receiver.go

```diff
@@ -13,9 +14,35 @@ type Receiver struct {
 // RiverCapsule marks receivers as a capsule.
 func (r Receiver) RiverCapsule() {}
 
-// FlowMetric is a wrapper around a single metric without the timestamp
+// FlowMetric is a wrapper around a single metric without the timestamp, FlowMetric is *mostly* immutable
+// the globalrefid can only be set if it is 0
 type FlowMetric struct {
-	GlobalRefID uint64
-	Labels      labels.Labels
-	Value       float64
+	globalRefID *atomic.Uint64
+	labels      labels.Labels
+	value       float64
+}
+
+// NewFlowMetric instantiates a new flow metric
+func NewFlowMetric(globalRefID uint64, lbls labels.Labels, value float64) *FlowMetric {
+	return &FlowMetric{
+		globalRefID: atomic.NewUint64(globalRefID),
+		labels:      lbls,
+		value:       value,
+	}
+}
+
+// GlobalRefID Retrieves the GlobalRefID
+func (fw *FlowMetric) GlobalRefID() uint64 { return fw.globalRefID.Load() }
+
+// SetGlobalRefID will only set the GlobalRefID if it is 0
+func (fw *FlowMetric) SetGlobalRefID(globalRefID uint64) {
+	fw.globalRefID.CAS(0, globalRefID)
+}
+
+// Value returns the value
+func (fw *FlowMetric) Value() float64 { return fw.value }
+
+// LabelsCopy returns a copy of the labels structure
+func (fw *FlowMetric) LabelsCopy() labels.Labels {
+	return fw.labels.Copy()
 }
```

Comment:
There should be a read-only method for labels. Making a copy every time we want to read will be expensive.

Reply:
When that use case arises, my plan was to add features on the FlowMetric itself. 
* ContainsLabel
* GetValue
etc

Reply:
If you're firm on copying the labels, can we write benchmarks for metrics.remote_write's receiver? 

We need to be cautious of changes that adds more overhead.

Reply:
When benching `GlobalRefMapping.GetGlobalRefIDByLabels(fms[j].labels)` vs `GlobalRefMapping.GetGlobalRefIDByLabels(fms[j].LabelsCopy())` the second is roughly 57% slower.

Reply:
Fiddling around to where the GlobalRefMapping grabs the underlying labels since it is in the same package and is safe to use in this specific instance.